libsrc/_DEVELOPMENT/math/float/math32/z80/f32_z80_mulu_de.asm

```diff
@@ -0,0 +1,117 @@
+;
+;  feilipu, 2019 May
+;
+;  This Source Code Form is subject to the terms of the Mozilla Public
+;  License, v. 2.0. If a copy of the MPL was not distributed with this
+;  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+;
+;-------------------------------------------------------------------------
+;
+; REPLICATION for Z80 of:
+; Z180 MLT DE and Z80-ZXN MUL DE
+;
+; unsigned multiplication of two 8-bit
+; multiplicands into a sixteen bit product
+;
+; enter : d = 8-bit multiplicand
+;         e = 8-bit multiplicand
+;
+; exit  : de = 16-bit product
+;         carry reset
+;
+; uses  : f, de
+   
+SECTION code_clib
+SECTION code_math
+
+PUBLIC m32_z80_mulu_de
+
+
+.m32_z80_mulu_de
+
+   inc d
+   dec d
+   jr Z,zero                ; multiply by 0
+
+   inc e
+   dec e
+   jr Z,zero                ; multiply by 0
+
+   push hl
+   ex de,hl
+
+   ld e,l
+   ld d,0
+
+   ; eliminate leading zero bits
+
+   sla h
+   jr C,branch_11
+
+   sla h
+   jr C,branch_12
+
+   sla h
+   jr C,branch_13
+
+   sla h
+   jr C,branch_14
+
+   sla h
+   jr C,branch_15
+
+   sla h
+   jr C,branch_16
+
+   sla h
+   jr C,branch_17
+
+   sla h
+   jr exit                ; multiply by 1
+   
+   ; multiplication tree
+
+.branch_11
+   add hl,hl
+   jr NC,branch_12
+   add hl,de
+
+.branch_12
+   add hl,hl
+   jr NC,branch_13
+   add hl,de
+
+.branch_13
+   add hl,hl
+   jr NC,branch_14
+   add hl,de
+
+.branch_14
+   add hl,hl
+   jr NC,branch_15
+   add hl,de
+
+.branch_15
+   add hl,hl
+   jr NC,branch_16
+   add hl,de
+
+.branch_16
+   add hl,hl
+   jr NC,branch_17
+   add hl,de
+
+.branch_17
+   add hl,hl
+   jr NC,exit
+   add hl,de
+
+.exit
+   ex de,hl
+   pop hl
+   or a
```

Comment:
why `or a`? if for carry flag, CF is already zero for all possible code paths here (after removing that last `sla h` above for D=1 case) (because either last `add hl,hl` was NC, or the following `add hl,de` did clear it)

Also if CF=0 is even important, then the `.zero` tail is not resetting it, so if any input is zero and CF=1, the routine will return DE=0, CF=1. But I think CF is not important at all?

Reply:
@ped7g thanks for taking the time to review. I appreciate it, and the feedback.

I was resetting carry, simply to align with the rest of the library code. I don't examine the carry at all. My thinking was muddy about whether this `mulu_de` was worth putting in the integer library, or not. Now, I believe it doesn't belong anywhere else. It is best to make it as fast as possible, since it is used exclusively by this library. It is the hottest code, so to speak. So, I'll remove `or a`.

Yes, as you say, neither the z80-zxn or z180 affect carry with their multiply instructions. You'll note that in the `z80_mulu_32_16x16` I have to capture the carry to preserve it past the last `mulu_de` for later use. This is not required in the z180 or z80-zxn versions of the code.